Disable ROOT automatic class parsing during finishSchedule()

fwyzard opened the pull request:
#### PR description:

Disable ROOT automatic class parsing during `ScheduleItems::finishSchedule()`, while checking for missing dictionaries.

Add a test to check that the framework handles correctly the case of `edm::Wrapper<T>` where:
  - `T` has a dictionary;
  - `edm::Wrapper<T>` does not have a dictionary;
  - the corresponding `classes.h` file includes CUDA headers.


#### PR validation:

Unit tests and HLT pass.

fwyzard committed:
Disable ROOT automatic class parsing while checking for missing dictionaries

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"054f633068adc1584e93c415802a9f5be3bba404":{"files":["FWCore/Framework/src/EventProcessor.cc"],"squashed":false,"time":1733258644}},"emoji":{},"last_seen_sha":"054f633068adc1584e93c415802a9f5be3bba404","signatures":{}} -->

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46858/42887

makortel commented:
Could you add e.g. "... during ScheduleItems::finishSchedule()` to the PR title? (to make it clear it the automatic parsing is still generally enabled)

cmsbuild commented:
A new Pull Request was created by  @fwyzard for master.

It involves the following packages:

- FWCore/Framework (**core**)


@Dr15Jones, @makortel, @smuzaffar can you please review it and eventually sign? Thanks.
@makortel, @missirol, @wddgit this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@wddgit @Dr15Jones Can you think of any drawbacks? (this part of the code should be fully serial)

fwyzard commented:
> Could you add e.g. "... during ScheduleItems::finishSchedule()` to the PR title? (to make it clear it the automatic parsing is still generally enabled)

Done, with only only finishSchedule() to stay on one line.

wddgit commented:
I cannot think of any problems this would cause. +1 from me.

cmsbuild commented:
+1

**Size**: This PR adds an extra 40KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-58643b/43228/summary.html
**COMMIT**: 054f633068adc1584e93c415802a9f5be3bba404
**CMSSW**: CMSSW_15_0_X_2024-12-03-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46858/43228/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-03-1100+58643b/65958/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3484682
 - DQMHistoTests: [Total failures: 382](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-03-1100+58643b/65958/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3484280
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 202 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-03-1100+58643b/65958/triggerResults): no differences found

Dr15Jones commented:
I'm concerned about placing the guard at that point. This decouples the setting of the behavior far from where the behavior needs to be changed. My suggestion is to

1. Change `ProductRegistry::setFrozen` to have a 3rd argument `bool disableRootAutoParsing` which is defaulted to `false`.
2. In `Schedule::finishSetup` pass `true` as the new argument to `ProductRegistry::setFrozen`
3. Add the guard conditionally inside `ProductRegistry::setFrozen` at the appropriate place

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46858/42958

cmsbuild commented:
-1

**Failed Tests**: RelVals 
**Size**: This PR adds an extra 60KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-58643b/43354/summary.html
**COMMIT**: 5d0a06568fdc892b71cbee1a13a5f2af1afaf03a
**CMSSW**: CMSSW_15_0_X_2024-12-09-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46858/43354/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


## RelVals

- 2024.202001```DAS Error```

fwyzard commented:
please test

fwyzard committed:
Add a test for missing dictionaries with CUDA headers

Add a test to check that the framework handles correctly the case of
edm::Wrapper<T> where:
  - T has a dictionary
  - edm::Wrapper<T> does not have a dictionary
  - the corresponding classes.h file includes CUDA headers

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46858/42964

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/BuildFile.xml  modified in PR(s): #46877
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #46877

cmsbuild commented:
Pull request #46858 was updated. @Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46858/42965

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Integration/plugins/BuildFile.xml  modified in PR(s): #46877
  - File  FWCore/Integration/test/BuildFile.xml  modified in PR(s): #46877

cmsbuild commented:
Pull request #46858 was updated. @Dr15Jones, @makortel, @smuzaffar can you please check and sign again.

fwyzard commented:
Without this PR, the new unit test fails with
```
 cmsRun FWCore/Integration/test/testMissingDictionaryCUDAChecking_cfg.py
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: could not acquire lock file for module 'cuda': failed to create unique file /cvmfs/cms.cern.ch/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm.lock-b9110f0d: Read-only file system [-Rmodule-build]
#include <cuda_runtime.h>
         ^
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: building module 'cuda' as '/cvmfs/cms.cern.ch/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm' [-Rmodule-build]
error: unable to open output file '/cvmfs/cms.cern.ch/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm': 'Read-only file system'
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: finished building module 'cuda' [-Rmodule-build]
DataFormatsTestObjects_x1r dictionary payload:59:10: fatal error: could not build module 'cuda'
#include <cuda_runtime.h>
 ~~~~~~~~^
----- Begin Fatal Exception 10-Dec-2024 17:51:30 CET-----------------------
An exception of category 'FatalRootError' occurred while
   [0] Constructing the EventProcessor
   [1] Calling edm::TypeWithDict::byName()
   Additional Info:
      [a] Getting TClass for edm::Wrapper<edmtest::MissingDictionaryCUDAObject>
      [b] Fatal Root Error: @SUB=TInterpreter::AutoParse
Error parsing payload code for class edmtest::MissingDictionaryCUDAObject with content:

#line 1 "DataFormatsTestObjects_x1r dictionary payload"

#ifndef CMS_DICT_IMPL
  #define CMS_DICT_IMPL 1
#endif
#ifndef _REENTRANT
  #define _REENTRANT 1
#endif
#ifndef GNUSOURCE
  #define GNUSOURCE 1
#endif
#ifndef __STRICT_ANSI__
  #define __STRICT_ANSI__ 1
#endif
#ifndef GNU_GCC
  #define GNU_GCC 1
#endif
#ifndef _GNU_SOURCE
  #define _GNU_SOURCE 1
#endif
#ifndef TBB_USE_GLIBCXX_VERSION
  #define TBB_USE_GLIBCXX_VERSION 120301
#endif
#ifndef TBB_SUPPRESS_DEPRECATED_MESSAGES
  #define TBB_SUPPRESS_DEPRECATED_MESSAGES 1
#endif
#ifndef TBB_PREVIEW_RESUMABLE_TASKS
  #define TBB_PREVIEW_RESUMABLE_TASKS 1
#endif
#ifndef TBB_PREVIEW_TASK_GROUP_EXTENSIONS
  #define TBB_PREVIEW_TASK_GROUP_EXTENSIONS 1
#endif
#ifndef BOOST_SPIRIT_THREADSAFE
  #define BOOST_SPIRIT_THREADSAFE 1
#endif
#ifndef PHOENIX_THREADSAFE
  #define PHOENIX_THREADSAFE 1
#endif
#ifndef BOOST_MATH_DISABLE_STD_FPCLASSIFY
  #define BOOST_MATH_DISABLE_STD_FPCLASSIFY 1
#endif
#ifndef BOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX
  #define BOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX 1
#endif
#ifndef CMSSW_GIT_HASH
  #define CMSSW_GIT_HASH "CMSSW_14_2_0_pre4"
#endif
#ifndef PROJECT_NAME
  #define PROJECT_NAME "CMSSW"
#endif
#ifndef PROJECT_VERSION
  #define PROJECT_VERSION "CMSSW_14_2_0_pre4"
#endif
#ifndef CMSSW_REFLEX_DICT
  #define CMSSW_REFLEX_DICT 1
#endif

#define _BACKWARD_BACKWARD_WARNING_H
// Inline headers
#include <cuda_runtime.h>

#include "DataFormats/Common/interface/Wrapper.h"
#include "DataFormats/TestObjects/interface/MissingDictionaryCUDAObject.h"

#undef  _BACKWARD_BACKWARD_WARNING_H


----- End Fatal Exception -------------------------------------------------
```

fwyzard commented:
With this PR we get a nicer error message:
```
$ cmsRun FWCore/Integration/test/testMissingDictionaryCUDAChecking_cfg.py
----- Begin Fatal Exception 10-Dec-2024 17:56:20 CET-----------------------
An exception of category 'DictionaryNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Calling ProductRegistry::initializeLookupTables, checking dictionaries for consumed products
Exception Message:
No data dictionary found for the following classes:

  edm::Wrapper<edmtest::MissingDictionaryCUDAObject>

Most likely each dictionary was never generated, but it may
be that it was generated in the wrong package. Please add
(or move) the specification '<class name="whatever"/>' to
the appropriate classes_def.xml file along with any other
information needed there. For example, if this class has any
transient members, you need to specify them in classes_def.xml.
Also include the class header in classes.h

The list of types above was generated while checking for
dictionaries related to products declared to be consumed.
A type listed above might or might not be a type declared
to be consumed. Instead it might be the type of a data member,
base class, wrapped type or other object needed by a consumed
type.  Below is some additional information which lists
the types declared to be consumed by a module and which
are associated with the types whose dictionaries were not
found:

  edmtest::MissingDictionaryCUDAObject

----- End Fatal Exception -------------------------------------------------
```
and the unit test passes.

cmsbuild commented:
+1

**Size**: This PR adds an extra 40KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-58643b/43360/summary.html
**COMMIT**: 1a82817b015bcac006f204f81b293ec7abf1ff1b
**CMSSW**: CMSSW_15_0_X_2024-12-10-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46858/43360/install.sh` to create a dev area with all the needed externals and cmssw changes.


* **DAS Queries**: The DAS query tests failed, see the summary page for details.


## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-10-1100+58643b/66058/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 10 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-10-1100+58643b/66058/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3510017
 - DQMHistoTests: [Total failures: 485](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-10-1100+58643b/66058/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3509512
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 202 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-10-1100+58643b/66058/triggerResults): no differences found

makortel commented on DataFormats/TestObjects/BuildFile.xml:
```diff
@@ -2,6 +2,15 @@
+  <flags LCG_DICT_HEADER="classes_cuda.h"/>
+  <flags LCG_DICT_XML="classes_cuda_def.xml"/>
+</iftool>
+
```
It seems to me the test (which itself is great!) doesn't really need the CUDA dependence, and the necessary and sufficient component of the test is the omission of the dictionary for `edm::Wrapper<edmtest::MissingDictionaryCUDAObject>`. Without CUDA dependence, and with the auto-parsing enabled, the test job would then succeed, after which the test itself would fail.

If I understood correctly, could you remove the dependence on CUDA (including names) from `DataFormats/TestObjects` and `FWCore/Integration`?

If you want to have a CUDA-specific test, it could be placed e.g. in `HeterogeneousCore/CUDATest`.

fwyzard replied:
If I comment out
```c++
//#include <cuda_runtime.h>
```
in `DataFormats/TestObjects/src/classes_cuda.h` and run the test _without the changes_ to disable the ROOT autoloading, I already get the "nice" message:
```
$ cmsRun FWCore/Integration/test/testMissingDictionaryCUDAChecking_cfg.py 
----- Begin Fatal Exception 18-Dec-2024 21:10:30 CET-----------------------
An exception of category 'DictionaryNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Calling ProductRegistry::initializeLookupTables, checking dictionaries for consumed products
Exception Message:
No data dictionary found for the following classes:

  edm::Wrapper<edmtest::MissingDictionaryCUDAObject>

Most likely each dictionary was never generated, but it may
be that it was generated in the wrong package. Please add
(or move) the specification '<class name="whatever"/>' to
the appropriate classes_def.xml file along with any other
information needed there. For example, if this class has any
transient members, you need to specify them in classes_def.xml.
Also include the class header in classes.h

The list of types above was generated while checking for
dictionaries related to products declared to be consumed.
A type listed above might or might not be a type declared
to be consumed. Instead it might be the type of a data member,
base class, wrapped type or other object needed by a consumed
type.  Below is some additional information which lists
the types declared to be consumed by a module and which
are associated with the types whose dictionaries were not
found:

  edmtest::MissingDictionaryCUDAObject

----- End Fatal Exception -------------------------------------------------
```

Which is the behaviour that we already get for missing dictionaries, normally.
So, without the CUDA include, the new test already succeeds without this PR.

It is the CUDA include (potentially other externals, too - I used CUDA because this is how we noticed the issue in the first place) that makes ROOT fail with a different exception
```
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: could not acquire lock file for module 'cuda': failed to create unique file /data/cmssw/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm.lock-1d54d871: Permission denied [-Rmodule-build]
#include <cuda_runtime.h>
         ^
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: building module 'cuda' as '/data/cmssw/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm' [-Rmodule-build]
error: unable to open output file '/data/cmssw/el8_amd64_gcc12/lcg/root/6.30.09-0fd387670244508fd2077d7feaf8b1c8/lib/cuda.pcm': 'Permission denied'
DataFormatsTestObjects_x1r dictionary payload:59:10: remark: finished building module 'cuda' [-Rmodule-build]
DataFormatsTestObjects_x1r dictionary payload:59:10: fatal error: could not build module 'cuda'
#include <cuda_runtime.h>
 ~~~~~~~~^
----- Begin Fatal Exception 18-Dec-2024 21:15:26 CET-----------------------
An exception of category 'FatalRootError' occurred while
   [0] Constructing the EventProcessor
   [1] Calling edm::TypeWithDict::byName()
   Additional Info:
      [a] Getting TClass for edm::Wrapper<edmtest::MissingDictionaryCUDAObject>
      [b] Fatal Root Error: @SUB=TInterpreter::AutoParse
Error parsing payload code for class edmtest::MissingDictionaryCUDAObject with content:
...
```
which prevents the framework from showing the "nice" message.

fwyzard replied:
In fact, I tried with a few other externals (zlib, boost, tbb) and none of them caused the weird exception that we get from CUDA.

fwyzard replied:
> If you want to have a CUDA-specific test, it could be placed e.g. in `HeterogeneousCore/CUDATest`.

Sure, I can move the test if you prefer.

What about the test data formats ?

makortel replied:
> in `DataFormats/TestObjects/src/classes_cuda.h` and run the test _without the changes_ to disable the ROOT autoloading, I already get the "nice" message:

Darn :( I'm confused why that happens though. I _guess_ in that case ROOT succeeds in the auto parsing itself, but the auto parsing does not lead to a dictionary for `edm::Wrapper<...>` (according to https://root.cern.ch/doc/master/classTClass.html information "From a lookup in the AST built by cling" "does not carry the information which is necessary to ROOT to perform I/O of instances of the class/struct it represents", so maybe we catch that).

We have one existing test that fails if ROOT ends up auto parsing the headers [`IOPool/Input/test/testNoParentDictionary.sh`](https://github.com/cms-sw/cmssw/blob/master/IOPool/Input/test/testNoParentDictionary.sh), but the setup is quite heavy. If the pattern would be applied here, I'd expect something simpler to be sufficient. Maybe adjusting `ROOT_INCLUDE_PATH
https://github.com/cms-sw/cmssw/blob/666efe2ff1988126cebcb0a3aaad35b8d4d0312a/IOPool/Input/test/testNoParentDictionary.sh#L42
and making the relevant header to fail to compile
https://github.com/cms-sw/cmssw/blob/666efe2ff1988126cebcb0a3aaad35b8d4d0312a/IOPool/Input/test/testNoParentDictionary.sh#L49
would be enough.


> > If you want to have a CUDA-specific test, it could be placed e.g. in `HeterogeneousCore/CUDATest`.
> 
> Sure, I can move the test if you prefer.

Thank you.

> What about the test data formats ?

Given that they have a dependence on CUDA (even if only via dictionary compilation), they should be placed outside of `DataFormats/TestObjects` as well. I see we already have a transient test data format defined in `HeterogeneousCore/CUDATest`, so I'd think these test data formats could be moved there as well.

makortel commented on FWCore/Integration/test/testMissingDictionaryCUDAChecking.sh:
```diff
@@ -0,0 +1,13 @@
+
+LOCAL_TEST_DIR=${SCRAM_TEST_PATH:-$CMSSW_BASE/src/FWCore/Integration/test}
+
+cmsRun ${LOCAL_TEST_DIR}/testMissingDictionaryCUDAChecking_cfg.py >& testMissingDictionaryCUDAChecking.log
+grep -q "An exception of category 'DictionaryNotFound' occurred" testMissingDictionaryCUDAChecking.log || die "Cannot find the following string in the exception message:\nAn exception of category 'DictionaryNotFound' occurred" $?
```
I'd add a check here that the `cmsRun` exited with a non-zero exit code.

fwyzard replied:
Sure.

fwyzard commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46858/43107

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #46929
  - File  FWCore/Integration/plugins/BuildFile.xml  modified in PR(s): #46929

cmsbuild commented:
Pull request #46858 was updated. @Dr15Jones, @makortel, @smuzaffar can you please check and sign again.

cmsbuild commented:
-1

**Failed Tests**: UnitTests RelVals-INPUT 
**Size**: This PR adds an extra 40KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-58643b/43555/summary.html
**COMMIT**: ac20e4bfa16d9505a5d9bf6fde0e2dd8e6787ffb
**CMSSW**: CMSSW_15_0_X_2024-12-18-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46858/43555/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIntegrationMissingDictionaryCUDA had ERRORS
</pre>

## RelVals-INPUT

- 2024.303001```DAS Error```
- 2024.302001```DAS Error```
- 2024.201001```DAS Error```
<details>
<summary>Expand to see more relval errors ...</summary>

- 2024.103001
- 2024.300001
</details>



## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 5 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-18-1100+58643b/66231/validateJR.html)
 - DQMHistoTests: Total files compared: 46
 - DQMHistoTests: Total histograms compared: 3510017
 - DQMHistoTests: [Total failures: 430](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-18-1100+58643b/66231/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3509567
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 45 files compared)
 - Checked 202 log files, 172 edm output root files, 46 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2024-12-18-1100+58643b/66231/triggerResults): no differences found

fwyzard commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-58643b/44075/summary.html
**COMMIT**: ac20e4bfa16d9505a5d9bf6fde0e2dd8e6787ffb
**CMSSW**: CMSSW_15_0_X_2025-01-29-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/46858/44075/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIntegrationMissingDictionaryCUDA had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-29-2300+58643b/66678/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 42 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-29-2300+58643b/66678/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4016770
 - DQMHistoTests: [Total failures: 1306](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-29-2300+58643b/66678/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4015444
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 218 log files, 189 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_0_X_2025-01-29-2300+58643b/66678/triggerResults): no differences found

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_15_1_X. Please open a backport if it should also go in to CMSSW_15_0_X.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-58643b/44457/summary.html
**COMMIT**: ac20e4bfa16d9505a5d9bf6fde0e2dd8e6787ffb
**CMSSW**: CMSSW_15_1_X_2025-02-17-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46858/44457/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIntegrationMissingDictionaryCUDA had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - [Reco comparison results: 5 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-2300+58643b/66972/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3905432
 - DQMHistoTests: [Total failures: 67](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-2300+58643b/66972/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3905345
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-17-2300+58643b/66972/triggerResults): no differences found

fwyzard commented:
Coming back to this, the [new test](https://github.com/cms-sw/cmssw/blob/ac20e4bfa16d9505a5d9bf6fde0e2dd8e6787ffb/FWCore/Integration/test/testMissingDictionaryCUDAChecking.sh) actually shows that the error
```
DataFormatsTestObjects_x1r dictionary payload:62:10: remark: could not acquire lock file for module 'cuda': failed to create unique file /cvmfs/cms.cern.ch/el8_amd64_gcc12/lcg/root/6.32.09-0a945d6e24dcaabe218b38fa8292785d/lib/cuda.pcm.lock-5561d58c: Read-only file system
      [-Rmodule-build]
#include <cuda_runtime.h>
         ^
DataFormatsTestObjects_x1r dictionary payload:62:10: remark: building module 'cuda' as '/cvmfs/cms.cern.ch/el8_amd64_gcc12/lcg/root/6.32.09-0a945d6e24dcaabe218b38fa8292785d/lib/cuda.pcm' [-Rmodule-build]
error: unable to open output file '/cvmfs/cms.cern.ch/el8_amd64_gcc12/lcg/root/6.32.09-0a945d6e24dcaabe218b38fa8292785d/lib/cuda.pcm': 'Read-only file system'
DataFormatsTestObjects_x1r dictionary payload:62:10: remark: finished building module 'cuda' [-Rmodule-build]
DataFormatsTestObjects_x1r dictionary payload:62:10: fatal error: could not build module 'cuda'
#include <cuda_runtime.h>
 ~~~~~~~~^
```
that leads to the fatal exception actually arises in the `edm::Schedule` constructor, rather than in `edm::Schedule::finishSetup`:
```
Thread 1 "cmsRun" hit Catchpoint 1 (exception thrown), 0x00007ffff712b2f1 in __cxxabiv1::__cxa_throw (obj=0x7fffcc4eae00, tinfo=0x7ffff7ea9690 <typeinfo for edm::Exception>, 
    dest=0x7ffff7e6f4a0 <edm::Exception::~Exception()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:81
81      ../../../../libstdc++-v3/libsupc++/eh_throw.cc: No such file or directory.
(gdb) bt
#0  0x00007ffff712b2f1 in __cxxabiv1::__cxa_throw (obj=0x7fffcc4eae00, tinfo=0x7ffff7ea9690 <typeinfo for edm::Exception>, dest=0x7ffff7e6f4a0 <edm::Exception::~Exception()>)
    at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:81
#1  0x00007ffff1c63b39 in (anonymous namespace)::RootErrorHandlerImpl(int, char const*, char const*) [clone .cold] ()
   from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#2  0x00007ffff784d374 in ErrorHandler () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCore.so
#3  0x00007ffff779a5e4 in TObject::Error(char const*, char const*, ...) const () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCore.so
#4  0x00007fffe9cd305c in TCling::AutoParseImplRecurse(char const*, bool) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCling.so
#5  0x00007fffe9cdd8ec in TCling::AutoParse(char const*) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCling.so
#6  0x00007fffe9cb048a in TClingLookupHelper__AutoParse(char const*) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCling.so
...
#25 0x00007ffff7c9d738 in edm::StreamSchedule::StreamSchedule(std::shared_ptr<edm::TriggerResultInserter>, std::vector<edm::propagate_const<std::shared_ptr<edm::PathStatusInserter> >, std::allocator<edm::propagate_const<std::shared_ptr<edm::PathStatusInserter> > > >&, std::vector<edm::propagate_const<std::shared_ptr<edm::EndPathStatusInserter> >, std::allocator<edm::propagate_const<std::shared_ptr<edm::EndPathStatusInserter> > > >&, std::shared_ptr<edm::ModuleRegistry>, edm::ParameterSet&, edm::service::TriggerNamesService const&, edm::PreallocationConfiguration const&, edm::SignallingProductRegistry&, edm::ExceptionToActionTable const&, std::shared_ptr<edm::ActivityRegistry>, std::shared_ptr<edm::ProcessConfiguration const>, edm::StreamID, edm::ProcessContext const*) ()
   from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#26 0x00007ffff7c827a2 in edm::Schedule::Schedule(edm::ParameterSet&, edm::service::TriggerNamesService const&, edm::SignallingProductRegistry&, edm::ExceptionToActionTable const&, std::shared_ptr<edm::ActivityRegistry>, std::shared_ptr<edm::ProcessConfiguration const>, edm::PreallocationConfiguration const&, edm::ProcessContext const*, edm::ModuleTypeResolverMaker const*) ()
   from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#27 0x00007ffff7c96b51 in edm::ScheduleItems::initModules(edm::ParameterSet&, edm::service::TriggerNamesService const&, edm::PreallocationConfiguration const&, edm::ProcessContext const*, edm::ModuleTypeResolverMaker const*) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
```

fwyzard commented:
If I disable class parsing just around constructing the `StreamSchedule`
```diff
diff --git a/FWCore/Framework/src/Schedule.cc b/FWCore/Framework/src/Schedule.cc
index 8796f935138..4bede1c5cb6 100644
--- a/FWCore/Framework/src/Schedule.cc
+++ b/FWCore/Framework/src/Schedule.cc
@@ -527,6 +528,8 @@ namespace edm {
     assert(0 < prealloc.numberOfStreams());
     streamSchedules_.reserve(prealloc.numberOfStreams());
     for (unsigned int i = 0; i < prealloc.numberOfStreams(); ++i) {
+      // disable ROOT automatic class parsing while constructing the modules
+      edm::SetClassParsing guard(false);
       streamSchedules_.emplace_back(make_shared_noexcept_false<StreamSchedule>(resultsInserter(),
                                                                                pathStatusInserters_,
                                                                                endPathStatusInserters_,
```
 I get the expected behaviour, that is, a nice error message about the missing dictionary instead of the ROOT message:
```
----- Begin Fatal Exception 18-Feb-2025 23:14:13 CET-----------------------
An exception of category 'DictionaryNotFound' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=edmtest::MissingDictionaryCUDAProducer label='producer'
   [2] Calling ProductRegistryHelper::addToRegistry, checking dictionaries for produced types
Exception Message:
No data dictionary found for the following classes:

  edm::Wrapper<edmtest::MissingDictionaryCUDAObject>

Most likely each dictionary was never generated, but it may
be that it was generated in the wrong package. Please add
(or move) the specification '<class name="whatever"/>' to
the appropriate classes_def.xml file along with any other
information needed there. For example, if this class has any
transient members, you need to specify them in classes_def.xml.
Also include the class header in classes.h

A type listed above might or might not be the same as a
type declared by a producer module with the function 'produces'.
Instead it might be the type of a data member, base class,
wrapped type, or other object needed by a produced type. Below
is some additional information which lists the types declared
to be produced by a producer module that are associated with
the types whose dictionaries were not found:

  edmtest::MissingDictionaryCUDAObject

----- End Fatal Exception -------------------------------------------------
```

Is that a good place for it, or would it cause problems with some other features that actually need the class parsing ?

Dr15Jones commented:
I would think it would break any code using the `cutParser` or the like as they need the class parser to run in order to probe the interfaces of the classes.

fwyzard commented:
How about disabling the class parsing only inside `edm::ProducerBase::registerProducts()` ?

fwyzard commented:
Indeed, this seems to work as well:
```diff
diff --git a/FWCore/Framework/src/ProducerBase.cc b/FWCore/Framework/src/ProducerBase.cc
index 2b28284a83d..0efd433dc87 100644
--- a/FWCore/Framework/src/ProducerBase.cc
+++ b/FWCore/Framework/src/ProducerBase.cc
@@ -69,6 +69,9 @@ namespace edm {
     }
     TypeLabelList const& plist = typeLabelList();
 
+    // disable ROOT automatic class parsing while registering the products
+    edm::SetClassParsing guard(false);
+
     ProductRegistryHelper::addToRegistry(plist.begin(), plist.end(), md, *(iReg), this, isListener);
     if (registrationCallback()) {
       iReg->watchProductAdditions(CallbackWrapper(producer, registrationCallback(), iReg, md));
```

fwyzard commented:
Maybe even better inside `ProductRegistryHelper::addToRegistry()` ?

fwyzard commented:
Yes, that also works, and it seems as close to the point of failure as it makes sense to go:
```diff
diff --git a/FWCore/Framework/src/ProductRegistryHelper.cc b/FWCore/Framework/src/ProductRegistryHelper.cc
index 290cdccaac1..94df82ae265 100644
--- a/FWCore/Framework/src/ProductRegistryHelper.cc
+++ b/FWCore/Framework/src/ProductRegistryHelper.cc
@@ -41,6 +42,9 @@ namespace edm {
     std::vector<std::string> producedTypes;
     std::set<std::tuple<BranchType, std::type_index, std::string>> registeredProducts;
 
+    // disable ROOT automatic class parsing while registering the products
+    edm::SetClassParsing guard(false);
+
     for (TypeLabelList::const_iterator p = iBegin; p != iEnd; ++p) {
       if (p->transition_ == Transition::BeginRun && not iProd->hasAbilityToProduceInBeginRuns()) {
         throwProducesWithoutAbility("BeginRun", p->typeID_.userClassName());
```

fwyzard commented:
See https://github.com/cms-sw/cmssw/pull/47397 for an alternative approach.

fwyzard commented:
Rebased, and addressed the comments about the new test.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-46858/43781

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  HeterogeneousCore/CUDATest/interface/MissingDictionaryCUDAObject.h  modified in PR(s): #47397
  - File  HeterogeneousCore/CUDATest/plugins/MissingDictionaryCUDAProducer.cc  modified in PR(s): #47397
  - File  HeterogeneousCore/CUDATest/src/classes.h  modified in PR(s): #47397
  - File  HeterogeneousCore/CUDATest/src/classes_def.xml  modified in PR(s): #47397
  - File  HeterogeneousCore/CUDATest/test/BuildFile.xml  modified in PR(s): #47397
  - File  HeterogeneousCore/CUDATest/test/testMissingDictionaryCUDA.sh  modified in PR(s): #47397
  - File  HeterogeneousCore/CUDATest/test/testMissingDictionaryCUDA_cfg.py  modified in PR(s): #47397

cmsbuild commented:
Pull request #46858 was updated. @Dr15Jones, @cmsbuild, @fwyzard, @makortel, @smuzaffar can you please check and sign again.

fwyzard commented:
After more digging I think I found what is the underlying issue: https://github.com/cms-sw/cmssw/issues/47400 .

makortel commented:
@cmsbuild, please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-58643b/44547/summary.html
**COMMIT**: 568c8b13737ed32ea6783379188730bbeffb2423
**CMSSW**: CMSSW_15_1_X_2025-02-20-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/46858/44547/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test testMissingDictionaryCUDA had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 3 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-20-1100+58643b/67045/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 14 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-20-1100+58643b/67045/validateJR.html)
 - DQMHistoTests: Total files compared: 49
 - DQMHistoTests: Total histograms compared: 3920300
 - DQMHistoTests: [Total failures: 9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-20-1100+58643b/67045/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3920271
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 48 files compared)
 - Checked 214 log files, 184 edm output root files, 49 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-02-20-1100+58643b/67045/triggerResults): no differences found

fwyzard commented:
> Would you still happen to have the stack trace available between frames 6 and 25?

Here is the full stack trace from a fresh run:
```rust
Thread 1 "cmsRun" hit Catchpoint 1 (exception thrown), 0x00007ffff712c2f1 in __cxxabiv1::__cxa_throw (obj=0x7fffcc1b5780, tinfo=0x7ffff7ea9690 <typeinfo for edm::Exception>, dest=0x7ffff7e6f4a0 <edm::Exception::~Exception()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:81
81      ../../../../libstdc++-v3/libsupc++/eh_throw.cc: No such file or directory.
#0  0x00007ffff712c2f1 in __cxxabiv1::__cxa_throw (obj=0x7fffcc1b5780, tinfo=0x7ffff7ea9690 <typeinfo for edm::Exception>, dest=0x7ffff7e6f4a0 <edm::Exception::~Exception()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:81
#1  0x00007ffff1c63b39 in (anonymous namespace)::RootErrorHandlerImpl(int, char const*, char const*) [clone .cold] () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#2  0x00007ffff784e374 in ErrorHandler () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCore.so
#3  0x00007ffff779b5e4 in TObject::Error(char const*, char const*, ...) const () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCore.so
#4  0x00007fffe9c9405c in TCling::AutoParseImplRecurse(char const*, bool) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCling.so
#5  0x00007fffe9c9e8ec in TCling::AutoParse(char const*) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCling.so
#6  0x00007fffe9c7148a in TClingLookupHelper__AutoParse(char const*) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCling.so
#7  0x00007fffe9bcae37 in ROOT::TMetaUtils::TClingLookupHelper::GetPartiallyDesugaredNameWithScopeHandling(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, bool) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCling.so
#8  0x00007ffff7849fed in TClassEdit::GetNormalizedName(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::basic_string_view<char, std::char_traits<char> >) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCore.so
#9  0x00007ffff7868160 in TClass::GetClass(char const*, bool, bool, unsigned long, unsigned long) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/external/el8_amd64_gcc12/lib/libCore.so
#10 0x00007ffff6b29101 in edm::TypeWithDict::byName(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreReflection.so
#11 0x00007ffff6b2991f in edm::TypeWithDict::byName(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreReflection.so
#12 0x00007ffff6b7a2d0 in edm::BranchDescription::initFromDictionary() () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libDataFormatsProvenance.so
#13 0x00007ffff6b7a5a3 in edm::BranchDescription::BranchDescription(edm::BranchType const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, edm::TypeWithDict const&, bool, bool, std::set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) () from /cvmfs/cms.cern.ch/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libDataFormatsProvenance.so
#14 0x00007ffff7c5caae in edm::ProductRegistryHelper::addToRegistry(__gnu_cxx::__normal_iterator<edm::ProductRegistryHelper::TypeLabelItem const*, std::vector<edm::ProductRegistryHelper::TypeLabelItem, std::allocator<edm::ProductRegistryHelper::TypeLabelItem> > > const&, __gnu_cxx::__normal_iterator<edm::ProductRegistryHelper::TypeLabelItem const*, std::vector<edm::ProductRegistryHelper::TypeLabelItem, std::allocator<edm::ProductRegistryHelper::TypeLabelItem> > > const&, edm::ModuleDescription const&, edm::ProductRegistry&, edm::ProductRegistryHelper*, bool) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#15 0x00007ffff7c5fafe in edm::ProducerBase::registerProducts(edm::ProducerBase*, edm::SignallingProductRegistry*, edm::ModuleDescription const&) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#16 0x00007fffcea68cb8 in edm::maker::ModuleHolderT<edm::global::EDProducerBase>::registerProductsAndCallbacks(edm::SignallingProductRegistry*) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/pluginHeterogeneousCoreCUDATestPlugins.so
#17 0x00007ffff7cc0624 in edm::Maker::makeModule(edm::MakeModuleParams const&, edm::signalslot::Signal<void (edm::ModuleDescription const&)>&, edm::signalslot::Signal<void (edm::ModuleDescription const&)>&) const () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#18 0x00007ffff7c2a7d8 in edm::Factory::makeModule(edm::MakeModuleParams const&, edm::ModuleTypeResolverMaker const*, edm::signalslot::Signal<void (edm::ModuleDescription const&)>&, edm::signalslot::Signal<void (edm::ModuleDescription const&)>&) const () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#19 0x00007ffff7c3b23f in edm::ModuleRegistry::getModule(edm::MakeModuleParams const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, edm::signalslot::Signal<void (edm::ModuleDescription const&)>&, edm::signalslot::Signal<void (edm::ModuleDescription const&)>&) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#20 0x00007ffff7cc0dca in edm::WorkerRegistry::getWorker(edm::WorkerParams const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#21 0x00007ffff7cc10a5 in edm::WorkerManager::getWorker(edm::ParameterSet&, edm::SignallingProductRegistry&, edm::PreallocationConfiguration const*, std::shared_ptr<edm::ProcessConfiguration const>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#22 0x00007ffff7c9a5f0 in edm::(anonymous namespace)::getWorker(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, edm::ParameterSet&, edm::WorkerManager&, edm::SignallingProductRegistry&, edm::PreallocationConfiguration const*, std::shared_ptr<edm::ProcessConfiguration const>) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#23 0x00007ffff7ca313f in edm::StreamSchedule::fillWorkers(edm::ParameterSet&, edm::SignallingProductRegistry&, edm::PreallocationConfiguration const*, std::shared_ptr<edm::ProcessConfiguration const>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, std::vector<edm::WorkerInPath, std::allocator<edm::WorkerInPath> >&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, edm::ConditionalTaskHelper const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#24 0x00007ffff7ca9c3f in edm::StreamSchedule::fillTrigPath(edm::ParameterSet&, edm::SignallingProductRegistry&, edm::PreallocationConfiguration const*, std::shared_ptr<edm::ProcessConfiguration const>, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<edm::HLTGlobalStatus>, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, edm::ConditionalTaskHelper const&, std::unordered_set<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#25 0x00007ffff7c9e578 in edm::StreamSchedule::StreamSchedule(std::shared_ptr<edm::TriggerResultInserter>, std::vector<edm::propagate_const<std::shared_ptr<edm::PathStatusInserter> >, std::allocator<edm::propagate_const<std::shared_ptr<edm::PathStatusInserter> > > >&, std::vector<edm::propagate_const<std::shared_ptr<edm::EndPathStatusInserter> >, std::allocator<edm::propagate_const<std::shared_ptr<edm::EndPathStatusInserter> > > >&, std::shared_ptr<edm::ModuleRegistry>, edm::ParameterSet&, edm::service::TriggerNamesService const&, edm::PreallocationConfiguration const&, edm::SignallingProductRegistry&, edm::ExceptionToActionTable const&, std::shared_ptr<edm::ActivityRegistry>, std::shared_ptr<edm::ProcessConfiguration const>, edm::StreamID, edm::ProcessContext const*) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#26 0x00007ffff7c835da in edm::Schedule::Schedule(edm::ParameterSet&, edm::service::TriggerNamesService const&, edm::SignallingProductRegistry&, edm::ExceptionToActionTable const&, std::shared_ptr<edm::ActivityRegistry>, std::shared_ptr<edm::ProcessConfiguration const>, edm::PreallocationConfiguration const&, edm::ProcessContext const*, edm::ModuleTypeResolverMaker const*) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#27 0x00007ffff7c97991 in edm::ScheduleItems::initModules(edm::ParameterSet&, edm::service::TriggerNamesService const&, edm::PreallocationConfiguration const&, edm::ProcessContext const*, edm::ModuleTypeResolverMaker const*) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#28 0x00007ffff7c012a7 in tbb::detail::d1::function_task<edm::EventProcessor::init(std::shared_ptr<edm::ProcessDesc>&, edm::ServiceToken const&, edm::serviceregistry::ServiceLegacy)::{lambda()#1}>::execute(tbb::detail::d1::execution_data&) [clone .lto_priv.0] () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#29 0x00007ffff7e313e1 in tbb::detail::r1::task_dispatcher::local_wait_for_all<false, tbb::detail::r1::external_waiter> (waiter=..., t=<optimized out>, this=0x7ffff40c3e00) at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-9b0f135342cfe979b6500c59f501774e/tbb-v2021.9.0/src/tbb/task_dispatcher.h:322
#30 tbb::detail::r1::task_dispatcher::local_wait_for_all<tbb::detail::r1::external_waiter> (waiter=..., t=<optimized out>, this=0x7ffff40c3e00) at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-9b0f135342cfe979b6500c59f501774e/tbb-v2021.9.0/src/tbb/task_dispatcher.h:458
#31 tbb::detail::r1::task_dispatcher::execute_and_wait (t=<optimized out>, wait_ctx=..., w_ctx=...) at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-9b0f135342cfe979b6500c59f501774e/tbb-v2021.9.0/src/tbb/task_dispatcher.cpp:168
#32 0x00007ffff7bcd21f in edm::EventProcessor::init(std::shared_ptr<edm::ProcessDesc>&, edm::ServiceToken const&, edm::serviceregistry::ServiceLegacy) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#33 0x00007ffff7bd0718 in edm::EventProcessor::EventProcessor(std::shared_ptr<edm::ProcessDesc>, edm::ServiceToken const&, edm::serviceregistry::ServiceLegacy) () from /data/user/fwyzard/disable_ROOT_automatic_class_parsing/CMSSW_15_0_0_pre3/lib/el8_amd64_gcc12/libFWCoreFramework.so
#34 0x0000000000408369 in tbb::detail::d1::task_arena_function<main::{lambda()#1}::operator()() const::{lambda()#1}, void>::operator()() const ()
#35 0x00007ffff7e1d9ad in tbb::detail::r1::task_arena_impl::execute (ta=..., d=warning: RTTI symbol not found for class 'tbb::detail::d1::task_arena_function<main::{lambda()#1}::operator()() const::{lambda()#1}, void>'
...) at /data/cmsbld/jenkins/workspace/ib-run-pr-tests/testBuildDir/BUILD/el8_amd64_gcc12/external/tbb/v2021.9.0-9b0f135342cfe979b6500c59f501774e/tbb-v2021.9.0/src/tbb/arena.cpp:688
#36 0x000000000040a283 in main::{lambda()#1}::operator()() const ()
#37 0x00000000004051b8 in main ()
```

fwyzard commented:
Superseded by https://github.com/cms-sw/cmsdist/pull/9765
